File: packages/geometry-utils/src/place-flow.ts
import ClipperLib from 'js-clipper';
import { IPoint, IPolygon, PlacementWorkerData } from './types';
import { polygonArea, rotatePolygon } from './helpers';
import ClipperWrapper from './clipper-wrapper';
import { almostEqual, generateNFPCacheKey, toRotationIndex } from './shared-helpers';
import Point from './point';
import Polygon from './polygon';
import PointPool from './point-pool';
import { NFP_INFO_START_INDEX, NFP_SHIFT_AMOUNT } from './constants';

function fillPointMemSeg(
    pointPool: PointPool,
    memSeg: Float64Array,
    points: IPoint[],
    offset: IPoint,
    prevValue: number
): number {
    const pointIndices: number = pointPool.alloc(1);
    const tmpPoint: Point = pointPool.get(pointIndices, 0);
    const pointCount = points.length;
    let i: number = 0;

    for (i = 0; i < pointCount; ++i) {
        tmpPoint
            .update(points[i])
            .add(offset)
            .fill(memSeg, prevValue + i);
    }

    pointPool.malloc(pointIndices);

    return prevValue + pointCount;
}

function bindNFP(polygon: Polygon, memSeg: Float64Array, index: number): void {
    const compressedInfo: number = memSeg[NFP_INFO_START_INDEX + index];
    const offset: number = compressedInfo >>> NFP_SHIFT_AMOUNT;
    const size: number = (compressedInfo & ((1 << NFP_SHIFT_AMOUNT) - 1)) >>> 1;

    polygon.bind(memSeg, offset, size);
}

function applyNfps(polygon: Polygon, clipper: ClipperLib.Clipper, nfpBuffer: ArrayBuffer, offset: IPoint): void {
    const nfpMemSeg: Float64Array = new Float64Array(nfpBuffer);
    const nfpCount: number = nfpMemSeg[1];
    let clone: ClipperLib.IntPoint[] = null;
    let i: number = 0;

    for (i = 0; i < nfpCount; ++i) {
        bindNFP(polygon, nfpMemSeg, i);
        clone = ClipperWrapper.toClipper(polygon, 1, offset, false, ClipperWrapper.CLEAN_TRASHOLD);

        if (clone.length > 2 && Math.abs(ClipperLib.Clipper.Area(clone)) > ClipperWrapper.AREA_TRASHOLD) {
            clipper.AddPath(clone, ClipperLib.PolyType.ptSubject, true);
        }
    }
}

// ensure all necessary NFPs exist
function getNfpError(placementData: PlacementWorkerData, placed: IPolygon[], path: IPolygon): boolean {
    let i: number = 0;
    let key: number = 0;

    for (i = 0; i < placed.length; ++i) {
        key = generateNFPCacheKey(placementData.angleSplit, false, placed[i], path);

        if (!placementData.nfpCache.has(key)) {
            return true;
        }
    }

    return false;
}

function nfpToClipper(polygon: Polygon, pointPool: PointPool, nfpMmSeg: Float64Array): ClipperLib.IntPoint[][] {
    const pointIndices = pointPool.alloc(1);
    const offset: Point = pointPool.get(pointIndices, 0).set(0, 0);
    const nfpCount: number = nfpMmSeg[1];
    let i: number = 0;
    const result = [];

    for (i = 0; i < nfpCount; ++i) {
        bindNFP(polygon, nfpMmSeg, i);
        result.push(ClipperWrapper.toClipper(polygon, 1, offset, true));
    }

    pointPool.malloc(pointIndices);

    return result;
}

function getFinalNfps(
    polygon: Polygon,
    pointPool: PointPool,
    placementData: PlacementWorkerData,
    placed: IPolygon[],
    path: IPolygon,
    binNfp: Float64Array,
    placement: number[]
) {
    const pointIndices: number = pointPool.alloc(1);
    const tmpPoint: Point = pointPool.get(pointIndices, 0);
    let clipper = new ClipperLib.Clipper();
    let i: number = 0;
    let key: number = 0;

    for (i = 0; i < placed.length; ++i) {
        key = generateNFPCacheKey(placementData.angleSplit, false, placed[i], path);

        if (!placementData.nfpCache.has(key)) {
            continue;
        }

        tmpPoint.fromMemSeg(placement, i);

        applyNfps(polygon, clipper, placementData.nfpCache.get(key), tmpPoint);
    }

    pointPool.malloc(pointIndices);

    const combinedNfp = new ClipperLib.Paths();

    if (
        !clipper.Execute(
            ClipperLib.ClipType.ctUnion,
            combinedNfp,
            ClipperLib.PolyFillType.pftNonZero,
            ClipperLib.PolyFillType.pftNonZero
        )
    ) {
        return null;
    }

    // difference with bin polygon
    let finalNfp: ClipperLib.Paths = new ClipperLib.Paths();
    const clipperBinNfp: ClipperLib.IntPoint[][] = nfpToClipper(polygon, pointPool, binNfp);

    clipper = new ClipperLib.Clipper();
    clipper.AddPaths(combinedNfp, ClipperLib.PolyType.ptClip, true);
    clipper.AddPaths(clipperBinNfp, ClipperLib.PolyType.ptSubject, true);

    if (
        !clipper.Execute(
            ClipperLib.ClipType.ctDifference,
            finalNfp,
            ClipperLib.PolyFillType.pftNonZero,
            ClipperLib.PolyFillType.pftNonZero
        )
    ) {
        return null;
    }

    finalNfp = ClipperLib.Clipper.CleanPolygons(finalNfp, ClipperWrapper.CLEAN_TRASHOLD);

    for (i = 0; i < finalNfp.length; ++i) {
        if (finalNfp[i].length < 3 || Math.abs(ClipperLib.Clipper.Area(finalNfp[i])) < ClipperWrapper.AREA_TRASHOLD) {
            finalNfp.splice(i, 1);
            --i;
        }
    }

    return finalNfp.length === 0 ? null : finalNfp;
}

function getPathKey(id: number, rotation: number, placementData: PlacementWorkerData): number {
    return (id << 16) | toRotationIndex(rotation, placementData.angleSplit);
}

function getResult(placements: number[][], pathItems: number[][], fitness: number): Float64Array {
    const placementCount: number = pathItems.length;
    const info = new Float64Array(placementCount);
    let totalSize: number = NFP_INFO_START_INDEX + placementCount;
    let offset: number = 0;
    let size: number = 0;
    let i: number = 0;

    for (i = 0; i < placementCount; ++i) {
        size = pathItems[i].length;
        info[i] = size | (totalSize << NFP_SHIFT_AMOUNT);
        totalSize += size * 3;
    }

    const result = new Float64Array(totalSize);

    result[0] = fitness;
    result[1] = placementCount;

    result.set(info, NFP_INFO_START_INDEX);

    for (i = 0; i < placementCount; ++i) {
        offset = info[i] >>> NFP_SHIFT_AMOUNT;
        size = info[i] & ((1 << NFP_SHIFT_AMOUNT) - 1);
        result.set(pathItems[i], offset);
        result.set(placements[i], offset + size);
    }

    return result;
}

export function placePaths(inputPaths: IPolygon[], placementData: PlacementWorkerData, pointPool: PointPool): Float64Array {
    if (!placementData.binPolygon) {
        return new Float64Array(0);
    }

    // rotate paths by given rotation
    const paths: IPolygon[] = [];
    const emptyPath: IPolygon = [] as IPolygon;
    const pointIndices: number = pointPool.alloc(1);
    const tmpPoint: Point = pointPool.get(pointIndices, 0);
    let i: number = 0;
    let j: number = 0;
    let k: number = 0;
    let m: number = 0;
    let path: IPolygon = null;
    let rotatedPath: IPolygon = null;

    emptyPath.id = -1;
    emptyPath.rotation = 0;

    for (i = 0; i < inputPaths.length; ++i) {
        path = inputPaths[i];
        rotatedPath = rotatePolygon(path, path.rotation);
        rotatedPath.rotation = path.rotation;
        rotatedPath.source = path.source;
        rotatedPath.id = path.id;
        paths.push(rotatedPath);
    }

    const polygon: Polygon = Polygon.create();
    const placements: number[][] = [];
    let placement: number[] = [];
    const pathItems: number[][] = [];
    let pathItem: number[] = [];
    const area = Math.abs(polygonArea(placementData.binPolygon));
    const pntMemSeg: Float64Array = new Float64Array(8192);
    const nfpMemSeg: Float64Array = new Float64Array(2048);
    let positionX: number = 0;
    let positionY: number = 0;
    let fitness: number = 0;
    let pointCount: number = 0;
    let key: number = 0;
    let minWidth: number = 0;
    let curArea: number = 0;
    let placed: IPolygon[] = [];
    let binNfp: Float64Array = null;
    let finalNfp: ClipperLib.Paths = null;
    let minArea: number = 0;
    let minX: number = 0;
    let nfpSize: number = 0;
    let binNfpCount: number = 0;
    let pathKey: number = 0;

    while (paths.length > 0) {
        placed = [];
        placement = [];
        pathItem = [];
        ++fitness; // add 1 for each new bin opened (lower fitness is better)

        for (i = 0; i < paths.length; ++i) {
            path = paths[i];
            pathKey = getPathKey(path.id, path.rotation, placementData);

            // inner NFP
            key = generateNFPCacheKey(placementData.angleSplit, true, emptyPath, path);
            binNfp = placementData.nfpCache.has(key) ? new Float64Array(placementData.nfpCache.get(key)) : null;

            // part unplaceable, skip             part unplaceable, skip
            if (!binNfp || binNfp.length < 3 || getNfpError(placementData, placed, path)) {
                continue;
            }

            positionX = NaN;

            binNfpCount = binNfp[1];

            if (placed.length === 0) {
                // first placement, put it on the left
                for (j = 0; j < binNfpCount; ++j) {
                    bindNFP(polygon, binNfp, j);

                    for (k = 0; k < polygon.length; ++k) {
                        tmpPoint.update(polygon.at(k)).sub(path[0]);

                        if (Number.isNaN(positionX) || tmpPoint.x < positionX) {
                            positionX = tmpPoint.x;
                            positionY = tmpPoint.y;
                        }
                    }
                }

                pathItem.push(pathKey);
                placement.push(positionX);
                placement.push(positionY);
                placed.push(path);

                continue;
            }

            finalNfp = getFinalNfps(polygon, pointPool, placementData, placed, path, binNfp, placement);

            if (finalNfp === null) {
                continue;
            }

            // choose placement that results in the smallest bounding box
            // could use convex hull instead, but it can create oddly shaped nests (triangles or long slivers)
            // which are not optimal for real-world use
            // OLD-TODO generalize gravity direction
            minWidth = 0;
            minArea = NaN;
            minX = NaN;
            curArea = 0;

            for (j = 0; j < finalNfp.length; ++j) {
                nfpSize = finalNfp[j].length;
                ClipperWrapper.toMemSeg(finalNfp[j], nfpMemSeg);
                polygon.bind(nfpMemSeg, 0, nfpSize);

                if (Math.abs(polygon.area) < 2) {
                    continue;
                }

                for (k = 0; k < nfpSize; ++k) {
                    pointCount = 0;

                    for (m = 0; m < placed.length; ++m) {
                        tmpPoint.fromMemSeg(placement, m);
                        pointCount = fillPointMemSeg(pointPool, pntMemSeg, placed[m], tmpPoint, pointCount);
                    }

                    polygon.bind(nfpMemSeg, 0, nfpSize);

                    tmpPoint.update(polygon.at(k)).sub(path[0]);

                    pointCount = fillPointMemSeg(pointPool, pntMemSeg, path, tmpPoint, pointCount);

                    polygon.bind(pntMemSeg, 0, pointCount);
                    // weigh width more, to help compress in direction of gravity
                    curArea = polygon.size.x * 2 + polygon.size.y;

                    if (
                        Number.isNaN(minArea) ||
                        curArea < minArea ||
                        (almostEqual(minArea, curArea) && (Number.isNaN(minX) || tmpPoint.x < minX))
                    ) {
                        minArea = curArea;
                        minWidth = polygon.size.x;
                        positionX = tmpPoint.x;
                        positionY = tmpPoint.y;
                        minX = tmpPoint.x;
                    }
                }
            }

            if (!Number.isNaN(positionX)) {
                placed.push(path);
                pathItem.push(pathKey);
                placement.push(positionX);
                placement.push(positionY);
            }
        }

        if (minWidth) {
            fitness += minWidth / area;
        }

        for (i = 0; i < placed.length; ++i) {
            const index = paths.indexOf(placed[i]);
            if (index !== -1) {
                paths.splice(index, 1);
            }
        }

        if (placement.length === 0) {
            break; // something went wrong
        }

        placements.push(placement);
        pathItems.push(pathItem);
    }

    // there were parts that couldn't be placed
    fitness += paths.length << 1;

    pointPool.malloc(pointIndices);

    return getResult(placements, pathItems, fitness);
}

File: packages/wasm-nesting/src/nesting/place_flow.rs
use crate::clipper::constants::CLIPPER_SCALE;
use crate::clipper::utils::{apply_nfps, get_combined_nfps, get_final_nfp};
use crate::geometry::point::Point;
use crate::nesting::nfp_wrapper::NFPWrapper;
use crate::nesting::place_content::PlaceContent;
use crate::nesting::polygon_node::PolygonNode;
use crate::utils::almost_equal::AlmostEqual;
use crate::utils::bit_ops::join_u16;
use crate::utils::math::write_uint32_to_f32;
use crate::utils::number::Number;

const NFP_INFO_START_INDEX: usize = 2;

/// Convert Int32Array polygon points to f32 memory segment, scaling down by CLIPPER_SCALE
fn to_mem_seg(polygon: &[Point<i32>]) -> Vec<f32> {
    let point_count = polygon.len();
    let mut result = vec![0.0f32; point_count << 1];
    let scale = CLIPPER_SCALE as f32;

    for i in 0..point_count {
        result[i << 1] = polygon[i].x as f32 / scale;
        result[(i << 1) + 1] = polygon[i].y as f32 / scale;
    }

    result
}

/// Fill point memory segment from a polygon node with an offset
fn fill_point_mem_seg(node: &PolygonNode, offset: &Point<f32>) -> Vec<f32> {
    let mut result = Vec::new();
    let point_count = node.mem_seg.len() >> 1;

    for i in 0..point_count {
        result.push(node.mem_seg[i << 1] + offset.x);
        result.push(node.mem_seg[(i << 1) + 1] + offset.y);
    }

    result
}

/// Get first placement position by finding leftmost point in bin NFP
/// Takes byte buffer (ArrayBuffer representation) instead of f32 buffer
pub fn get_first_placement(nfp_buffer: &[u8], first_point: &Point<f32>) -> Vec<f32> {
    let mut position_x = f32::NAN;
    let mut position_y = f32::NAN;

    // Read NFP count from buffer (little-endian u32 at byte offset 4)
    if nfp_buffer.len() < 8 {
        return vec![position_x, position_y];
    }

    let nfp_count =
        u32::from_le_bytes([nfp_buffer[4], nfp_buffer[5], nfp_buffer[6], nfp_buffer[7]]) as usize;

    let mut byte_offset = NFP_INFO_START_INDEX * 4; // Convert to byte offset

    for _ in 0..nfp_count {
        if byte_offset + 4 > nfp_buffer.len() {
            break;
        }

        // Read compressed info (offset and size)
        let compressed_info = u32::from_le_bytes([
            nfp_buffer[byte_offset],
            nfp_buffer[byte_offset + 1],
            nfp_buffer[byte_offset + 2],
            nfp_buffer[byte_offset + 3],
        ]);
        byte_offset += 4;

        use crate::utils::bit_ops::get_u16;
        let data_offset = get_u16(compressed_info, 1) as usize * 4; // Convert to byte offset
        let size = get_u16(compressed_info, 0) as usize;
        let point_count = size >> 1;

        if data_offset + size * 4 > nfp_buffer.len() {
            continue;
        }

        for j in 0..point_count {
            let x_offset = data_offset + (j << 1) * 4;
            let y_offset = x_offset + 4;

            if y_offset + 4 > nfp_buffer.len() {
                break;
            }

            let x = f32::from_le_bytes([
                nfp_buffer[x_offset],
                nfp_buffer[x_offset + 1],
                nfp_buffer[x_offset + 2],
                nfp_buffer[x_offset + 3],
            ]) - first_point.x;

            let y = f32::from_le_bytes([
                nfp_buffer[y_offset],
                nfp_buffer[y_offset + 1],
                nfp_buffer[y_offset + 2],
                nfp_buffer[y_offset + 3],
            ]) - first_point.y;

            if position_x.is_nan() || x < position_x {
                position_x = x;
                position_y = y;
            }
        }
    }

    vec![position_x, position_y]
}

pub fn get_placement_data(
    final_nfp: &Vec<Vec<Point<i32>>>,
    placed: &[PolygonNode],
    node: &PolygonNode,
    placement: &[f32],
    first_point: &Point<f32>,
    input_y: f32,
) -> Vec<f32> {
    let mut position_x = f32::NAN;
    let mut position_y = input_y;
    let mut min_width = 0.0f32;
    let mut min_area = f32::NAN;
    let mut min_x = f32::NAN;
    let mut cur_area: f32;
    let mut tmp_point = Point::<f32>::new(None, None);

    for j in 0..final_nfp.len() {
        let nfp_size = final_nfp[j].len();
        let mem_seg1 = to_mem_seg(&final_nfp[j]);

        if f32::abs_polygon_area(&mem_seg1) < 2.0 {
            continue;
        }

        for k in 0..nfp_size {
            let mut buffer = Vec::new();

            for m in 0..placed.len() {
                tmp_point.x = placement[m << 1];
                tmp_point.y = placement[(m << 1) + 1];
                buffer.extend(fill_point_mem_seg(&placed[m], &tmp_point));
            }

            tmp_point.x = mem_seg1[k << 1] - first_point.x;
            tmp_point.y = mem_seg1[(k << 1) + 1] - first_point.y;

            buffer.extend(fill_point_mem_seg(node, &tmp_point));

            let mem_seg2 = buffer.into_boxed_slice();
            let point_count = mem_seg2.len() >> 1;

            let bounds = f32::calculate_bounds(&mem_seg2, 0, point_count);
            // weigh width more, to help compress in direction of gravity
            cur_area = bounds[2] * 2.0 + bounds[3];

            if min_area.is_nan()
                || cur_area < min_area
                || (min_area.almost_equal(cur_area, Some(f32::tol()))
                    && (min_x.is_nan() || tmp_point.x < min_x))
            {
                min_area = cur_area;
                min_width = bounds[2];
                position_x = tmp_point.x;
                position_y = tmp_point.y;
                min_x = tmp_point.x;
            }
        }
    }

    if !position_x.is_nan() {
        vec![min_width, position_y, position_x]
    } else {
        vec![min_width, position_y]
    }
}

/// Convert NFPWrapper to clipper format (Vec<Vec<Point<i32>>>)
fn nfp_to_clipper(nfp_wrapper: &NFPWrapper) -> Vec<Vec<Point<i32>>> {
    let nfp_count = nfp_wrapper.count();
    let mut result = Vec::with_capacity(nfp_count);

    for i in 0..nfp_count {
        let mem_seg = nfp_wrapper.get_nfp_mem_seg(i);
        let point_count = mem_seg.len() >> 1;
        let mut polygon = Vec::with_capacity(point_count);

        for j in 0..point_count {
            let x = (mem_seg[j << 1] * CLIPPER_SCALE as f32).round() as i32;
            let y = (mem_seg[(j << 1) + 1] * CLIPPER_SCALE as f32).round() as i32;
            polygon.push(Point { x, y });
        }

        result.push(polygon);
    }

    result
}

/// Get final NFPs by combining placed nodes' NFPs and subtracting from bin NFP
/// Rust port of ClipperWrapper.getFinalNfps
pub fn get_final_nfps(
    place_content: &PlaceContent,
    placed: &[PolygonNode],
    path: &PolygonNode,
    bin_nfp: &NFPWrapper,
    placement: &[f32],
) -> Vec<Vec<Point<i32>>> {
    let mut total_nfps: Vec<Vec<Point<i32>>> = Vec::new();

    // Collect and apply all NFPs from placed nodes
    for (i, placed_node) in placed.iter().enumerate() {
        let key = PolygonNode::generate_nfp_cache_key(
            place_content.rotations(),
            false,
            placed_node,
            path,
        );

        if let Some(nfp_cache_value) = place_content.nfp_cache().get(&key) {
            // Convert Vec<u8> to Vec<f32>
            let f32_count = nfp_cache_value.len() / 4;
            let mut f32_buffer = Vec::with_capacity(f32_count);

            for j in 0..f32_count {
                let offset = j * 4;
                f32_buffer.push(f32::from_le_bytes([
                    nfp_cache_value[offset],
                    nfp_cache_value[offset + 1],
                    nfp_cache_value[offset + 2],
                    nfp_cache_value[offset + 3],
                ]));
            }

            // Get offset point from placement
            let offset_x = placement[i << 1];
            let offset_y = placement[(i << 1) + 1];
            let offset = Point {
                x: offset_x,
                y: offset_y,
            };

            // Apply NFPs with offset and add to total
            let applied = apply_nfps(f32_buffer, &offset);
            total_nfps.extend(applied);
        }
    }

    // Combine all NFPs using union
    let combined_nfp = get_combined_nfps(&total_nfps);

    if combined_nfp.is_empty() {
        return Vec::new();
    }

    // Convert bin NFP to clipper format
    let clipper_bin_nfp = nfp_to_clipper(bin_nfp);

    // Get final NFP by subtracting combined NFP from bin NFP
    get_final_nfp(&combined_nfp, &clipper_bin_nfp)
}

pub fn get_result(placements: &[Vec<f32>], path_items: &[Vec<u32>], fitness: f32) -> Vec<f32> {
    let placement_count = path_items.len();
    let mut info = vec![0u32; placement_count];
    let mut total_size = NFP_INFO_START_INDEX + placement_count;
    let mut merged_size: u32;
    let mut offset: usize;
    let mut size: usize;

    for i in 0..placement_count {
        size = path_items[i].len();
        merged_size = join_u16(size as u16, total_size as u16);
        info[i] = merged_size;
        total_size += size * 3;
    }

    let mut result = vec![0.0f32; total_size];

    result[0] = fitness;
    result[1] = placement_count as f32;

    for i in 0..placement_count {
        merged_size = info[i];
        offset = (merged_size >> 16) as usize;
        size = (merged_size & 0xFFFF) as usize;

        write_uint32_to_f32(&mut result, NFP_INFO_START_INDEX + i, merged_size);

        for j in 0..size {
            write_uint32_to_f32(&mut result, offset + j, path_items[i][j]);
        }

        result[offset + size..offset + size + placements[i].len()].copy_from_slice(&placements[i]);
    }

    result
}

/// Port of TypeScript placePaths function
/// Main placement algorithm that packs polygons into bins
pub fn place_paths(buffer: &[u8]) -> Vec<f32> {
    let mut place_content = PlaceContent::new();
    place_content.init(buffer);

    let mut first_point = Point::<f32> { x: 0.0, y: 0.0 };
    let mut placements: Vec<Vec<f32>> = Vec::new();
    let mut path_items: Vec<Vec<u32>> = Vec::new();
    let mut placement: Vec<f32> = Vec::new();
    let mut path_item: Vec<u32> = Vec::new();
    let mut position_y: f32 = 0.0;
    let mut fitness: f32 = 0.0;
    let mut min_width: f32 = 0.0;
    let mut placed: Vec<PolygonNode> = Vec::new();

    while place_content.node_count() > 0 {
        placed.clear();
        placement.clear();
        path_item.clear();
        fitness += 1.0; // add 1 for each new bin opened (lower fitness is better)

        let node_count = place_content.node_count();

        for i in 0..node_count {
            let node = place_content.node_at(i).clone();

            // Get first point from node's mem_seg
            if node.mem_seg.len() >= 2 {
                first_point.x = node.mem_seg[0];
                first_point.y = node.mem_seg[1];
            }

            let path_key = place_content.get_path_key(i);

            // Get bin NFP
            let bin_nfp_option = place_content.get_bin_nfp(i);
            if bin_nfp_option.is_none() {
                continue;
            }

            let bin_nfp_bytes = bin_nfp_option.unwrap();
            let bin_nfp = NFPWrapper::new(bin_nfp_bytes);

            // Part unplaceable, skip
            if bin_nfp.is_broken() || place_content.get_nfp_error(&placed, &node) {
                continue;
            }

            if placed.is_empty() {
                let placement_data = get_first_placement(bin_nfp_bytes, &first_point);

                if placement_data.len() >= 2 {
                    position_y = placement_data[1];

                    path_item.push(path_key);
                    placement.push(placement_data[0]);
                    placement.push(placement_data[1]);
                    placed.push(node.clone());
                }

                continue;
            }

            // Get final NFP
            let final_nfp = get_final_nfps(&place_content, &placed, &node, &bin_nfp, &placement);

            if final_nfp.is_empty() {
                continue;
            }

            // Get placement data
            let placement_data = get_placement_data(
                &final_nfp,
                &placed,
                &node,
                &placement,
                &first_point,
                position_y,
            );

            min_width = placement_data[0];
            position_y = placement_data[1];

            if placement_data.len() == 3 {
                placed.push(node.clone());
                path_item.push(path_key);
                placement.push(placement_data[2]);
                placement.push(placement_data[1]);
            }
        }

        if min_width != 0.0 {
            fitness += min_width / place_content.area();
        }

        // Remove placed nodes
        for placed_node in &placed {
            place_content.remove_node(placed_node);
        }

        if placement.is_empty() {
            break; // something went wrong
        }

        placements.push(placement.clone());
        path_items.push(path_item.clone());
    }

    // There were parts that couldn't be placed
    fitness += (place_content.node_count() << 1) as f32;

    get_result(&placements, &path_items, fitness)
}
